src/precompiles/ecmul.rs

```diff
@@ -0,0 +1,357 @@
+        } else {
+            let mut write_location = MemoryLocation {
+                memory_type: MemoryType::Heap, // we default for some value, here it's not that important
+                page: MemoryPage(params.memory_page_to_write),
+                index: MemoryIndex(params.output_memory_offset),
+            };
+
+            let err_marker = U256::zero();
+            let ok_or_err_query = MemoryQuery {
+                timestamp: timestamp_to_write,
+                location: write_location,
+                value: err_marker,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            let ok_or_err_query =
+                memory.execute_partial_query(monotonic_cycle_counter, ok_or_err_query);
+
+            write_location.index.0 += 1;
+            let empty_result = U256::zero();
+            let x_result_query = MemoryQuery {
+                timestamp: timestamp_to_write,
+                location: write_location,
+                value: empty_result,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            let x_result_query = memory.execute_partial_query(monotonic_cycle_counter, x_result_query);
+
+            write_location.index.0 += 1;
+            let empty_result = U256::zero();
+            let y_result_query = MemoryQuery {
+                timestamp: timestamp_to_write,
+                location: write_location,
+                value: empty_result,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            let y_result_query = memory.execute_partial_query(monotonic_cycle_counter, y_result_query);
+
+            if B {
+                round_witness.writes[0] = ok_or_err_query;
+                round_witness.writes[1] = x_result_query;
+                round_witness.writes[2] = y_result_query;
+                write_history.push(ok_or_err_query);
+                write_history.push(x_result_query);
+                write_history.push(y_result_query);
+            }
+        }
+
+        let witness = if B {
+            Some((read_history, write_history, vec![round_witness]))
+        } else {
+            None
+        };
+
+        (NUM_ROUNDS, witness)
+    }
+}
+
+/// This function multiplies the point (x1,y1) by the scalar s on the BN254 curve.
+/// It returns the result as a G1Affine point.
+///
+/// If the points are not on the curve or coordinates are not valid field elements,
+/// the function will return an error.
+pub fn ecmul_inner((x1, y1): (U256, U256), s: U256) -> Result<G1Affine> {
+    // Converting coordinates to the finite field format
+    // and validating that the conversion is successful
+    let x1_field = Fq::from_str(x1.to_string().as_str()).ok_or(Error::msg("invalid x1"))?;
+    let y1_field = Fq::from_str(y1.to_string().as_str()).ok_or(Error::msg("invalid y1"))?;
+    let s_field = Fr::from_str(s.to_string().as_str()).ok_or(Error::msg("invalid s"))?;
```

Comment:
Why is that? `s` can be an arbitrary uint256 number

Reply:
True, missed that, we will update

Reply:
Yeah, indeed, missed that. Now, `s` can be arbitrary. Also, I added three different tests where the scalar is greater than the group order to make sure we have not missed anything